autograder-api/src/main/java/de/firemage/autograder/api/loader/AutograderLoader.java

```diff
@@ -0,0 +1,114 @@
+package de.firemage.autograder.api.loader;
+
+import de.firemage.autograder.api.Linter;
+import de.firemage.autograder.api.TempLocation;
+import org.reflections.Reflections;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.net.URLConnection;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Path;
+import java.nio.file.Files;
+
+public class AutograderLoader {
+    private static final String AUTOGRADER_DOWNLOAD_URL = "https://github.com/Feuermagier/autograder/releases/latest/download/autograder-cmd.jar";
+    private static final String AUTOGRADER_RELEASE_PATH = "https://github.com/Feuermagier/autograder/releases/latest";
+
+    private static final Logger LOG = LoggerFactory.getLogger(AutograderLoader.class);
+
+    private static URLClassLoader autograderClassLoader = null;
+    private static String currentTag = null;
+    private static Path downloadedAutograderPath = null;
+
+    public static void loadFromFile(Path autograderPath) throws IOException {
+        loadAutograderIntoProcess(autograderPath);
+    }
+
+    public static void loadFromGithubWithExtraChecks() throws IOException {
+        String tag = getAutograderVersionTag();
+        if (downloadedAutograderPath == null || !downloadedAutograderPath.getFileName().startsWith(tag)) {
+            downloadAutograderRelease(tag);
+        }
+        currentTag = tag;
+        loadAutograderIntoProcess(downloadedAutograderPath);
+    }
+
+    public static boolean isCurrentVersionLoaded() throws IOException {
+        if (downloadedAutograderPath == null) {
+            return true;
+        }
+
+        return getAutograderVersionTag().equals(currentTag);
+    }
+
+    public static boolean isAutograderLoaded() {
+        return autograderClassLoader != null;
+    }
+
+    public static Linter instantiateLinter(Linter.Builder builder) {
+        return new ImplementationBinder<>(Linter.class)
+                .param(Linter.Builder.class, builder)
+                .classLoader(autograderClassLoader)
+                .instantiate();
+    }
+
+    public static TempLocation instantiateTempLocation(Path path) {
+        return new ImplementationBinder<>(TempLocation.class)
+                .param(Path.class, path)
+                .classLoader(autograderClassLoader)
+                .instantiate();
+    }
+
+    public static TempLocation instantiateTempLocation() {
+        try {
+            Class.forName("de.firemage.autograder.core.file.TempLocationImpl", true, autograderClassLoader);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
What will that do? And why do you handle it different from something like Linter?

Reply:
That was a leftover from debugging